From a005d37f48a405395764fd831145e02aeb424850 Mon Sep 17 00:00:00 2001
From: Fred Underwood <underwoodfred@proton.me>
Date: Mon, 23 Dec 2024 12:48:35 +1000
Subject: [PATCH] add second factor PIN preference to FingerprintSettings

Ported from GrapheneOS 31acf583f9. AXP fork: this tree has no fingerprint keyguard toggle, so the preference is set up in addFingerprintPreferences(), the controller is created for every sensor type and assigned in buildPreferenceControllers().
---
 res/values/strings.xml                        |   7 +
 res/xml/security_settings_fingerprint.xml     |   5 +
 ...icSecondFactorPinPreferenceController.java | 138 ++++++++++++++++++
 .../fingerprint/FingerprintSettings.java      | 124 +++++++++++++++-
 .../FingerprintSettingsFragmentTest.java      |  17 +++
 .../shadow/ShadowLockPatternUtils.java        |  14 +-
 6 files changed, 300 insertions(+), 5 deletions(-)
 create mode 100644 src/com/android/settings/biometrics/fingerprint/BiometricSecondFactorPinPreferenceController.java

diff --git a/res/values/strings.xml b/res/values/strings.xml
index a4e1c3c7..ba6c0ad2 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -994,6 +994,13 @@
     <!-- Positive button of the confirmation dialog displayed to the user when they attempt to delete biometric second factor PIN -->
     <string name="biometric_second_factor_pin_delete_confirmation_dialog_positive_button">Delete</string>
 
+    <!-- Title of preference within fingerprint settings for biometric second factor PIN -->
+    <string name="biometric_second_factor_pin_title">Second factor PIN</string>
+    <!-- Summary of preference within fingerprint settings for biometric second factor PIN when setup is needed -->
+    <string name="biometric_second_factor_pin_setup_summary">Setup needed</string>
+    <!-- Summary of preference within fingerprint settings for biometric second factor PIN when it has been setup -->
+    <string name="biometric_second_factor_pin_empty_summary"></string>
+
     <!-- Introduction detail message shown in fingerprint enrollment introduction screen in setup wizard. [CHAR LIMIT=NONE]-->
     <string name="security_settings_fingerprint_v2_enroll_introduction_message_setup">When you see this icon, use your fingerprint for authentication, like when you sign in to apps or approve a purchase</string>
     <!-- Introduction subtitle message shown in fingerprint enrollment introduction screen in setup wizard. [CHAR LIMIT=NONE]-->
diff --git a/res/xml/security_settings_fingerprint.xml b/res/xml/security_settings_fingerprint.xml
index 32f09244..0da0bbf1 100644
--- a/res/xml/security_settings_fingerprint.xml
+++ b/res/xml/security_settings_fingerprint.xml
@@ -56,4 +56,9 @@
         android:key="security_settings_fingerprint_footer">
     </PreferenceCategory>
 
+        <androidx.preference.Preference
+            android:key="biometric_second_factor_pin"
+            android:title="@string/biometric_second_factor_pin_title" />
+
+
 </PreferenceScreen>
diff --git a/src/com/android/settings/biometrics/fingerprint/BiometricSecondFactorPinPreferenceController.java b/src/com/android/settings/biometrics/fingerprint/BiometricSecondFactorPinPreferenceController.java
new file mode 100644
index 00000000..d6ca973f
--- /dev/null
+++ b/src/com/android/settings/biometrics/fingerprint/BiometricSecondFactorPinPreferenceController.java
@@ -0,0 +1,138 @@
+package com.android.settings.biometrics.fingerprint;
+
+import static com.android.internal.widget.LockDomain.Secondary;
+
+import android.app.admin.DevicePolicyManager;
+import android.app.settings.SettingsEnums;
+import android.content.Context;
+import android.content.Intent;
+import android.text.TextUtils;
+
+import androidx.fragment.app.Fragment;
+import androidx.preference.Preference;
+
+import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockscreenCredential;
+import com.android.settings.R;
+import com.android.settings.core.BasePreferenceController;
+import com.android.settings.overlay.FeatureFactory;
+import com.android.settings.password.ChooseBiometricSecondFactorPin;
+import com.android.settings.security.screenlock.BiometricSecondFactorPinSettings;
+import com.android.settingslib.RestrictedLockUtils;
+import com.android.settingslib.RestrictedLockUtilsInternal;
+
+/**
+ * Controller for the biometric second factor PIN preference. If the user does not have a second
+ * factor PIN set, launches an Activity for creating one, otherwise launches the settings for the
+ * second factor PIN.
+ */
+public class BiometricSecondFactorPinPreferenceController extends BasePreferenceController {
+
+    private final LockPatternUtils mLockPatternUtils;
+
+    protected final Fragment mHost;
+
+    private int mUserId = -1;
+    private LockscreenCredential mPrimaryCredential;
+
+    private int mSettingsRequestCode;
+    private int mChooseRequestCode;
+
+    public BiometricSecondFactorPinPreferenceController(Context context, Fragment host, String key) {
+        super(context, key);
+
+        mHost = host;
+        mLockPatternUtils = FeatureFactory.getFeatureFactory()
+                .getSecurityFeatureProvider()
+                .getLockPatternUtils(context);
+
+    }
+
+    public void setUserId(int userId) {
+        mUserId = userId;
+    }
+    public void setPrimaryCredential(LockscreenCredential primaryCredential) {
+        mPrimaryCredential = primaryCredential;
+    }
+
+    public void setSettingsRequestCode(int value) {
+        mSettingsRequestCode = value;
+    }
+
+    public void setChooseRequestCode(int value) {
+        mChooseRequestCode = value;
+    }
+
+    @Override
+    public int getMetricsCategory() {
+        return SettingsEnums.FINGERPRINT;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        if (mLockPatternUtils.isSecure(mUserId, Secondary)) {
+            return mContext.getResources().getString(
+                    R.string.biometric_second_factor_pin_empty_summary);
+        }
+        return mContext.getResources().getString(
+                R.string.biometric_second_factor_pin_setup_summary);
+
+    }
+
+    @Override
+    public boolean handlePreferenceTreeClick(Preference preference) {
+        if (!TextUtils.equals(preference.getKey(), getPreferenceKey())) {
+            return super.handlePreferenceTreeClick(preference);
+        }
+
+        if (mLockPatternUtils.isSecure(mUserId, Secondary)) {
+            launchSettings();
+        } else {
+            launchChoose();
+        }
+
+        return true;
+    }
+
+    public void launchSettings() {
+        mHost.startActivityForResult(getSettingsIntent(), mSettingsRequestCode);
+    }
+
+    private Intent getSettingsIntent() {
+        BiometricSecondFactorPinSettings.IntentBuilder builder =
+                new BiometricSecondFactorPinSettings.IntentBuilder(mContext)
+                        .setSourceMetricsCateogry(getMetricsCategory())
+                        .setUserId(mUserId)
+                        .setPrimaryCredential(mPrimaryCredential);
+
+        return builder.build();
+    }
+
+    private void launchChoose() {
+        mHost.startActivityForResult(getChooseIntent(), mChooseRequestCode);
+    }
+
+    private Intent getChooseIntent() {
+        ChooseBiometricSecondFactorPin.IntentBuilder builder =
+                new ChooseBiometricSecondFactorPin.IntentBuilder(mContext)
+                        .setUserId(mUserId)
+                        .setPrimaryCredential(mPrimaryCredential);
+
+        return builder.build();
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        if (mLockPatternUtils.checkUserSupportsBiometricSecondFactor(mUserId, false)
+                && mLockPatternUtils.isBiometricKeyguardEnabled(mUserId)
+                && getRestrictingAdmin() == null) {
+            return AVAILABLE;
+        }
+        return DISABLED_FOR_USER;
+    }
+
+    protected RestrictedLockUtils.EnforcedAdmin getRestrictingAdmin() {
+        return RestrictedLockUtilsInternal.checkIfKeyguardFeaturesDisabled(
+                mContext, DevicePolicyManager.KEYGUARD_DISABLE_FINGERPRINT, mUserId);
+    }
+}
diff --git a/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java b/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java
index 0c2d7e0d..c6e755a2 100644
--- a/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java
+++ b/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java
@@ -42,6 +42,7 @@ import android.hardware.fingerprint.FingerprintManager;
 import android.hardware.fingerprint.FingerprintSensorPropertiesInternal;
 import android.os.Bundle;
 import android.os.Handler;
+import android.os.Looper;
 import android.os.UserHandle;
 import android.os.UserManager;
 import android.os.VibrationEffect;
@@ -67,6 +68,7 @@ import androidx.preference.PreferenceViewHolder;
 import androidx.preference.TwoStatePreference;
 
 import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockscreenCredential;
 import com.android.settings.R;
 import com.android.settings.SubSettings;
 import com.android.settings.Utils;
@@ -78,10 +80,12 @@ import com.android.settings.core.SettingsBaseActivity;
 import com.android.settings.core.instrumentation.InstrumentedDialogFragment;
 import com.android.settings.dashboard.DashboardFragment;
 import com.android.settings.overlay.FeatureFactory;
+import com.android.settings.password.ChooseBiometricSecondFactorPin;
 import com.android.settings.password.ChooseLockGeneric;
 import com.android.settings.password.ChooseLockSettingsHelper;
 import com.android.settings.password.ConfirmDeviceCredentialActivity;
 import com.android.settings.search.BaseSearchIndexProvider;
+import com.android.settings.security.screenlock.BiometricSecondFactorPinSettings;
 import com.android.settingslib.HelpUtils;
 import com.android.settingslib.RestrictedLockUtils;
 import com.android.settingslib.RestrictedLockUtils.EnforcedAdmin;
@@ -178,7 +182,7 @@ public class FingerprintSettings extends SubSettings {
                     @Override
                     public List<AbstractPreferenceController>
                             createPreferenceControllers(Context context) {
-                        return createThePreferenceControllers(context);
+                        return createThePreferenceControllers(context, null);
                     }
 
                     private boolean hasEnrolledFingerprints(Context context) {
@@ -192,7 +196,7 @@ public class FingerprintSettings extends SubSettings {
                 };
 
         private static List<AbstractPreferenceController> createThePreferenceControllers(Context
-                context) {
+                context, FingerprintSettingsFragment host) {
             final List<AbstractPreferenceController> controllers = new ArrayList<>();
             FingerprintManager manager = Utils.getFingerprintManagerOrNull(context);
             if (manager == null || !manager.isHardwareDetected()) {
@@ -219,6 +223,7 @@ public class FingerprintSettings extends SubSettings {
                                 context,
                                 KEY_REQUIRE_SCREEN_ON_TO_AUTH
                         ));
+
             } else if (screenOffUnlockUdfps()) {
                 controllers.add(
                         new FingerprintUnlockCategoryController(
@@ -233,6 +238,15 @@ public class FingerprintSettings extends SubSettings {
             }
             controllers.add(new FingerprintsEnrolledCategoryPreferenceController(context,
                     KEY_FINGERPRINTS_ENROLLED_CATEGORY));
+            BiometricSecondFactorPinPreferenceController c =
+                    new BiometricSecondFactorPinPreferenceController(context, host,
+                            KEY_BIOMETRIC_SECOND_FACTOR_PIN);
+            c.setSettingsRequestCode(
+                    BIOMETRIC_SECOND_FACTOR_PIN_SETTINGS_REQUEST);
+            c.setChooseRequestCode(
+                    BIOMETRIC_SECOND_FACTOR_PIN_CHOOSE_REQUEST);
+            controllers.add(c);
+
             return controllers;
         }
 
@@ -253,6 +267,8 @@ public class FingerprintSettings extends SubSettings {
         private static final String KEY_LAUNCHED_CONFIRM = "launched_confirm";
         private static final String KEY_HAS_FIRST_ENROLLED = "has_first_enrolled";
         private static final String KEY_IS_ENROLLING = "is_enrolled";
+        private static final String KEY_LAUNCHED_BIOMETRIC_SECOND_FACTOR_PIN_ACTIVITY =
+                "launched_biometric_second_factor_pin_activity";
         @VisibleForTesting
         static final String KEY_REQUIRE_SCREEN_ON_TO_AUTH =
                 "security_settings_require_screen_on_to_auth";
@@ -267,6 +283,7 @@ public class FingerprintSettings extends SubSettings {
                 "security_settings_fingerprint_footer";
         private static final String KEY_BIOMETRICS_AUTHENTICATION_REQUESTED =
                 "biometrics_authentication_requested";
+        private static final String KEY_BIOMETRIC_SECOND_FACTOR_PIN = "biometric_second_factor_pin";
 
         private static final int MSG_REFRESH_FINGERPRINT_TEMPLATES = 1000;
         private static final int MSG_FINGER_AUTH_SUCCESS = 1001;
@@ -278,6 +295,8 @@ public class FingerprintSettings extends SubSettings {
         private static final int CONFIRM_REQUEST = 101;
         @VisibleForTesting
         static final int CHOOSE_LOCK_GENERIC_REQUEST = 102;
+        public static final int BIOMETRIC_SECOND_FACTOR_PIN_SETTINGS_REQUEST = 9999;
+        public static final int BIOMETRIC_SECOND_FACTOR_PIN_CHOOSE_REQUEST = 9998;
         @VisibleForTesting
         static final int ADD_FINGERPRINT_REQUEST = 10;
         private static final int AUTO_ADD_FIRST_FINGERPRINT_REQUEST = 11;
@@ -287,6 +306,8 @@ public class FingerprintSettings extends SubSettings {
         private List<AbstractPreferenceController> mControllers;
         private FingerprintUnlockCategoryController
                 mFingerprintUnlockCategoryPreferenceController;
+        private BiometricSecondFactorPinPreferenceController
+                mBiometricSecondFactorPinPreferenceController;
         private FingerprintSettingsRequireScreenOnToAuthPreferenceController
                 mRequireScreenOnToAuthPreferenceController;
         private FingerprintSettingsScreenOffUnlockUdfpsPreferenceController
@@ -299,12 +320,14 @@ public class FingerprintSettings extends SubSettings {
         private PreferenceCategory mFingerprintUnlockFooter;
         private boolean mFingerprintWakeAndUnlock;
         private boolean mProximityCheckOnFingerprintUnlock;
+        private Preference mBiometricSecondFactorPin;
 
         private FingerprintManager mFingerprintManager;
         private FingerprintUpdater mFingerprintUpdater;
         private List<FingerprintSensorPropertiesInternal> mSensorProperties;
         private boolean mInFingerprintLockout;
         private byte[] mToken;
+        private LockscreenCredential mUserPassword;
         private boolean mLaunchedConfirm;
         private boolean mBiometricsAuthenticationRequested;
         private boolean mHasFirstEnrolled = true;
@@ -312,6 +335,7 @@ public class FingerprintSettings extends SubSettings {
         private int mUserId;
         private final List<FooterColumn> mFooterColumns = new ArrayList<>();
         private boolean mIsEnrolling;
+        private boolean mLaunchedBiometricSecondFactorPinActivity;
 
         private long mChallenge;
 
@@ -528,6 +552,15 @@ public class FingerprintSettings extends SubSettings {
                         mHasFirstEnrolled);
                 mBiometricsAuthenticationRequested = savedInstanceState.getBoolean(
                         KEY_BIOMETRICS_AUTHENTICATION_REQUESTED);
+                mLaunchedBiometricSecondFactorPinActivity = savedInstanceState.getBoolean(
+                        KEY_LAUNCHED_BIOMETRIC_SECOND_FACTOR_PIN_ACTIVITY
+                );
+                mUserPassword = savedInstanceState.getParcelable(
+                        ChooseLockSettingsHelper.EXTRA_KEY_PASSWORD);
+                mBiometricSecondFactorPinPreferenceController.setPrimaryCredential(mUserPassword);
+                // This controller needs userId set in #onActivityResult, but upstream doesn't set
+                // them until #onResume.
+                mBiometricSecondFactorPinPreferenceController.setUserId(mUserId);
             }
 
             // (mLaunchedConfirm or mIsEnrolling) means that we are waiting an activity result.
@@ -684,9 +717,18 @@ public class FingerprintSettings extends SubSettings {
                     ((FingerprintSettingsPreferenceController) controller).setUserId(mUserId);
                 } else if (controller instanceof FingerprintUnlockCategoryController) {
                     ((FingerprintUnlockCategoryController) controller).setUserId(mUserId);
+                } else if (controller instanceof BiometricSecondFactorPinPreferenceController c) {
+                    c.setUserId(mUserId);
                 }
             }
 
+            // AXP fork: GrapheneOS sets this up next to its fingerprint keyguard toggle, which
+            // this tree doesn't have
+            mBiometricSecondFactorPin = findPreference(KEY_BIOMETRIC_SECOND_FACTOR_PIN);
+            if (mBiometricSecondFactorPin != null) {
+                updateSecondFactorPinVisibility();
+            }
+
             // This needs to be after setting ids, otherwise
             // |mRequireScreenOnToAuthPreferenceController.isChecked| is always checking the primary
             // user instead of the user with |mUserId|.
@@ -766,6 +808,20 @@ public class FingerprintSettings extends SubSettings {
             }
         }
 
+        private void updateSecondFactorPinVisibility() {
+            final boolean available =
+                    mBiometricSecondFactorPinPreferenceController.isAvailable();
+            if (mBiometricSecondFactorPin.isVisible() != available) {
+                mBiometricSecondFactorPin.setVisible(available);
+                if (available) {
+                    // Without this there will be no summary when preference was initially not
+                    // available.
+                    mBiometricSecondFactorPinPreferenceController.updateState(
+                            mBiometricSecondFactorPin);
+                }
+            }
+        }
+
         private void updateFingerprintUnlockCategoryVisibility() {
             final boolean fingerprintUnlockCategoryAvailable =
                     mFingerprintUnlockCategoryPreferenceController.isAvailable();
@@ -915,7 +971,8 @@ public class FingerprintSettings extends SubSettings {
         @Override
         public void onStop() {
             super.onStop();
-            if (!getActivity().isChangingConfigurations() && !mLaunchedConfirm && !mIsEnrolling) {
+            if (!getActivity().isChangingConfigurations() && !mLaunchedConfirm && !mIsEnrolling &&
+                    !mLaunchedBiometricSecondFactorPinActivity) {
                 setResult(RESULT_TIMEOUT);
                 getActivity().finish();
             }
@@ -941,6 +998,10 @@ public class FingerprintSettings extends SubSettings {
             outState.putBoolean(KEY_HAS_FIRST_ENROLLED, mHasFirstEnrolled);
             outState.putBoolean(KEY_BIOMETRICS_AUTHENTICATION_REQUESTED,
                     mBiometricsAuthenticationRequested);
+            outState.putBoolean(KEY_LAUNCHED_BIOMETRIC_SECOND_FACTOR_PIN_ACTIVITY,
+                    mLaunchedBiometricSecondFactorPinActivity);
+            outState.putParcelable(ChooseLockSettingsHelper.EXTRA_KEY_PASSWORD,
+                    mUserPassword != null ? mUserPassword.duplicate() : null);
         }
 
         @Override
@@ -961,6 +1022,8 @@ public class FingerprintSettings extends SubSettings {
                 FingerprintPreference fpref = (FingerprintPreference) pref;
                 final Fingerprint fp = fpref.getFingerprint();
                 showRenameDialog(fp);
+            } else if (KEY_BIOMETRIC_SECOND_FACTOR_PIN.equals(key)) {
+                mLaunchedBiometricSecondFactorPinActivity = true;
             }
             return super.onPreferenceTreeClick(pref);
         }
@@ -1044,7 +1107,14 @@ public class FingerprintSettings extends SubSettings {
 
         private List<AbstractPreferenceController> buildPreferenceControllers(Context context) {
             final List<AbstractPreferenceController> controllers =
-                    createThePreferenceControllers(context);
+                    createThePreferenceControllers(context, this);
+            if (controllers != null) {
+                for (AbstractPreferenceController controller : controllers) {
+                    if (controller instanceof BiometricSecondFactorPinPreferenceController c) {
+                        mBiometricSecondFactorPinPreferenceController = c;
+                    }
+                }
+            }
             if (!isUdfps() && context.getResources().getBoolean(
                     org.lineageos.platform.internal.R.bool.config_fingerprintWakeAndUnlock)) {
                 for (AbstractPreferenceController controller : controllers) {
@@ -1081,6 +1151,16 @@ public class FingerprintSettings extends SubSettings {
                 mLaunchedConfirm = false;
                 if (resultCode == RESULT_FINISHED || resultCode == RESULT_OK) {
                     if (BiometricUtils.containsGatekeeperPasswordHandle(data)) {
+                        if (data.hasExtra(ChooseLockSettingsHelper.EXTRA_KEY_PASSWORD)) {
+                            mUserPassword = data.getParcelableExtra(
+                                    ChooseLockSettingsHelper.EXTRA_KEY_PASSWORD);
+                            mBiometricSecondFactorPinPreferenceController.setPrimaryCredential(
+                                    mUserPassword);
+                        } else {
+                            Log.d(TAG, "Current primary credential missing");
+                            finish();
+                        }
+
                         if (!mHasFirstEnrolled && !mIsEnrolling) {
                             final Activity activity = getActivity();
                             if (activity != null) {
@@ -1194,6 +1274,32 @@ public class FingerprintSettings extends SubSettings {
                         finish();
                     }
                 }
+            } else if (requestCode == BIOMETRIC_SECOND_FACTOR_PIN_CHOOSE_REQUEST) {
+                mLaunchedBiometricSecondFactorPinActivity = false;
+                // This technique of relying on the launched Activity to indicate timeout isn't
+                // perfect. If an app calls finish() for something other than RESULT_TIMEOUT, but
+                // the app is in background, then on returning to foreground there will be no
+                // RESULT_TIMEOUT. This is easily seen by adding a sleep(1000ms) before the finish()
+                // call and going to background during this time.
+                // The technique is used inconsistently throughout Settings app, such as
+                // ChooseLockGeneric using it, but not ChooseLockPassword. However, it is used
+                // consistently within FingerprintSettings, so I use it for second factor.
+                if (resultCode == ChooseBiometricSecondFactorPin
+                        .ChooseBiometricSecondFactorPinFragment.RESULT_NOT_FOREGROUND) {
+                    getActivity().setResult(RESULT_TIMEOUT);
+                    getActivity().finish();
+                } else if (resultCode == ChooseBiometricSecondFactorPin
+                        .ChooseBiometricSecondFactorPinFragment.RESULT_FINISHED) {
+                    mLaunchedBiometricSecondFactorPinActivity = true;
+                    mBiometricSecondFactorPinPreferenceController.launchSettings();
+                }
+            } else if (requestCode == BIOMETRIC_SECOND_FACTOR_PIN_SETTINGS_REQUEST) {
+                mLaunchedBiometricSecondFactorPinActivity = false;
+                if (resultCode ==
+                        BiometricSecondFactorPinSettings.RESULT_NOT_FOREGROUND) {
+                    getActivity().setResult(RESULT_TIMEOUT);
+                    getActivity().finish();
+                }
             }
         }
 
@@ -1203,6 +1309,15 @@ public class FingerprintSettings extends SubSettings {
             if (getActivity().isFinishing()) {
                 mFingerprintManager.revokeChallenge(mUserId, mChallenge);
             }
+
+            if (mUserPassword != null) {
+                mUserPassword.zeroize();
+            }
+            new Handler(Looper.myLooper()).postDelayed(() -> {
+                System.gc();
+                System.runFinalization();
+                System.gc();
+            }, 5000);
         }
 
         private Drawable getHighlightDrawable() {
@@ -1261,6 +1376,7 @@ public class FingerprintSettings extends SubSettings {
                         true);
                 intent.putExtra(Intent.EXTRA_USER_ID, mUserId);
                 intent.putExtra(ChooseLockSettingsHelper.EXTRA_KEY_REQUEST_GK_PW_HANDLE, true);
+                intent.putExtra(ChooseLockSettingsHelper.EXTRA_KEY_RETURN_CREDENTIALS, true);
                 intent.putExtra(ChooseLockSettingsHelper.EXTRA_KEY_FOR_FINGERPRINT, true);
                 startActivityForResult(intent, CHOOSE_LOCK_GENERIC_REQUEST);
             }
diff --git a/tests/robotests/src/com/android/settings/biometrics/fingerprint/FingerprintSettingsFragmentTest.java b/tests/robotests/src/com/android/settings/biometrics/fingerprint/FingerprintSettingsFragmentTest.java
index 1086f85d..ba6caf71 100644
--- a/tests/robotests/src/com/android/settings/biometrics/fingerprint/FingerprintSettingsFragmentTest.java
+++ b/tests/robotests/src/com/android/settings/biometrics/fingerprint/FingerprintSettingsFragmentTest.java
@@ -69,6 +69,8 @@ import androidx.fragment.app.FragmentTransaction;
 import androidx.preference.Preference;
 import androidx.test.core.app.ApplicationProvider;
 
+import com.android.internal.widget.LockPatternUtils;
+import com.android.settings.biometrics.fingerprint.feature.SfpsRestToUnlockFeature;
 import com.android.settings.password.ChooseLockSettingsHelper;
 import com.android.settings.password.ConfirmDeviceCredentialActivity;
 import com.android.settings.search.BaseSearchIndexProvider;
@@ -137,6 +139,12 @@ public class FingerprintSettingsFragmentTest {
     private FingerprintAuthenticateSidecar mFingerprintAuthenticateSidecar;
     private FingerprintRemoveSidecar mFingerprintRemoveSidecar;
 
+    private FakeFeatureFactory mFeatureFactory;
+    @Mock
+    private LockPatternUtils mLockPatternUtils;
+    @Mock
+    private SfpsRestToUnlockFeature mSfpsRestToUnlockFeature;
+
     @Before
     public void setUp() {
         ShadowUtils.setFingerprintManager(mFingerprintManager);
@@ -151,6 +159,15 @@ public class FingerprintSettingsFragmentTest {
         when(mBiometricManager.canAuthenticate(PRIMARY_USER_ID,
                 BiometricManager.Authenticators.IDENTITY_CHECK))
                 .thenReturn(BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE);
+
+        // FeatureFactory mocking copied from ChangeScreenLockPreferenceControllerTest.
+        mFeatureFactory = FakeFeatureFactory.setupForTest();
+        when(mFeatureFactory.securityFeatureProvider.getLockPatternUtils(mContext))
+                .thenReturn(mLockPatternUtils);
+        when(mFeatureFactory.mFingerprintFeatureProvider.getSfpsRestToUnlockFeature(mContext))
+                .thenReturn(mSfpsRestToUnlockFeature);
+
+        when(mLockPatternUtils.checkUserSupportsBiometricSecondFactor(anyInt())).thenReturn(true);
     }
 
     @After
diff --git a/tests/robotests/testutils/com/android/settings/testutils/shadow/ShadowLockPatternUtils.java b/tests/robotests/testutils/com/android/settings/testutils/shadow/ShadowLockPatternUtils.java
index c5ad6c0a..55ad6cd6 100644
--- a/tests/robotests/testutils/com/android/settings/testutils/shadow/ShadowLockPatternUtils.java
+++ b/tests/robotests/testutils/com/android/settings/testutils/shadow/ShadowLockPatternUtils.java
@@ -16,6 +16,8 @@
 
 package com.android.settings.testutils.shadow;
 
+import static com.android.internal.widget.LockDomain.Primary;
+
 import android.app.admin.DevicePolicyManager;
 import android.app.admin.PasswordMetrics;
 import android.content.ComponentName;
@@ -26,6 +28,7 @@ import android.os.UserHandle;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
+import com.android.internal.widget.LockDomain;
 import com.android.internal.widget.LockPatternUtils;
 import com.android.internal.widget.LockscreenCredential;
 
@@ -51,6 +54,8 @@ public class ShadowLockPatternUtils {
     private static Map<Integer, Boolean> sUserToBiometricAllowedMap = new HashMap<>();
     private static Map<Integer, Boolean> sUserToLockPatternEnabledMap = new HashMap<>();
     private static Map<Integer, Integer> sKeyguardStoredPasswordQualityMap = new HashMap<>();
+    private static Map<Integer, Integer> sKeyguardStoredPasswordQualityMapSecondary =
+            new HashMap<>();
 
     private static boolean sIsUserOwnsFrpCredential;
 
@@ -99,7 +104,14 @@ public class ShadowLockPatternUtils {
 
     @Implementation
     protected int getKeyguardStoredPasswordQuality(int userHandle) {
-        return sKeyguardStoredPasswordQualityMap.getOrDefault(userHandle, /* defaultValue= */ 1);
+        return getKeyguardStoredPasswordQuality(userHandle, Primary);
+    }
+
+    @Implementation
+    protected int getKeyguardStoredPasswordQuality(int userHandle, LockDomain lockDomain) {
+        Map<Integer, Integer> passwordQualityMap = lockDomain == Primary ?
+                sKeyguardStoredPasswordQualityMap : sKeyguardStoredPasswordQualityMapSecondary;
+        return passwordQualityMap.getOrDefault(userHandle, /* defaultValue= */ 1);
     }
 
     @Implementation
-- 
2.43.0

